js/core/data_controller.js

```diff
@@ -0,0 +1,173 @@
+class DataController {
+    constructor(dataSource) {
+        this.updateDataSource(dataSource);
+    }
+
+    updateDataSource(dataSource) {
+        this._dataSource = dataSource;
+    }
+
+    load() {
+        return this._dataSource.load();
+    }
+
+    loadSingle(propName, propValue) {
+        if(arguments.length < 2) {
+            propValue = propName;
+            propName = this.key();
+        }
+
+        return this._dataSource.loadSingle(propName, propValue);
+    }
+
+    loadFromStore(loadOptions) {
+        return this.store().load(loadOptions);
+    }
+
+    loadDataSource() {
+        const dataSource = this._dataSource;
+
+        if(dataSource) {
+            if(dataSource.isLoaded()) {
+                this._proxiedDataSourceChangedHandler && this._proxiedDataSourceChangedHandler();
+            } else {
+                dataSource.load();
+            }
+        }
+    }
+
+    loadNextPage() {
+        this.pageIndex(1 + this.pageIndex());
+
+        return this.load();
+    }
+
+    loadOptions() {
+        return this._dataSource?.loadOptions();
+    }
+
+    userData() {
+        return this._dataSource?._userData;
+    }
+
+    cancel(operationId) {
+        this._dataSource?.cancel(operationId);
```

Comment:
let's use [null object pattern](https://sourcemaking.com/design_patterns/null_object) to get rid of all this optional chainings and conditions. You can find an example here: https://github.com/DevExpress/DevExtreme/blob/22_2/js/core/resize_observer.js